### PR TITLE
Metrics view tests

### PR DESCRIPTION
**Pull request by egor-ryashin**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,7 +26,7 @@ export/
 *.db
 *.wal
 *.parquet
-*.csv
+./*.csv
```
Comment (begelundmuller) [inline .gitignore:29]: This might have unintended side effects, consider including the specific path in the next section instead

File: runtime/server/instances.go [hunk 2]
```diff
@@ -49,7 +49,7 @@ func (s *Server) CreateInstance(ctx context.Context, req *runtimev1.CreateInstan
 	// Check OLAP connection
 	olap, err := drivers.Open(inst.OLAPDriver, inst.OLAPDSN)
 	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error()+" (Arguments: "+inst.OLAPDriver+" "+inst.OLAPDSN+")")
```
Comment (begelundmuller) [inline runtime/server/instances.go:52]: 1. Don't add a DSN to an error/log message, it could be bad for security
2. Make it a formatted error instead, like `status.Errorf(codes.InvalidArgument, "could not connect to driver '%s': %w", inst.OLAPDriver, err)`

File: runtime/server/metrics_views_test.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	server, err := NewServer(&ServerOptions{
+		ConnectionCacheSize: 100,
+	}, metastore, nil)
+	require.NoError(t, err)
+
+	resp, err := server.CreateInstance(context.Background(), &runtimev1.CreateInstanceRequest{
+		OlapDriver:   "duckdb",
+		OlapDsn:      "",
+		RepoDriver:   "file",
+		RepoDsn:      "../testproject/ad_bids",
```
Comment (begelundmuller) [inline runtime/server/metrics_views_test.go:30]: Name the test data folder `testdata` instead of `testproject` – it's a convention in go (e.g. see [here](https://dave.cheney.net/2016/05/10/test-fixtures-in-go))

**Review by begelundmuller — CHANGES_REQUESTED** [review]
This looks great – good setup with the test server with an adbids instance. Some other general comments:

1. In `runtime/testproject/ad_bids`, two rows feels a little fragile for the data. How about either using the real ~5MB `adbids.csv.gz` file or at least using a file with the first 1k rows to make sure the tests cover some real data variance.
2. Looks like `sf_props` is not used, so lets remove it from the test data. I think it's simpler anyway if all our tests build on just one test project that covers all kinds of use cases.

**Comment by egor-ryashin** [comment]
> In runtime/testproject/ad_bids, two rows feels a little fragile for the data. How about either using the real ~5MB adbids.csv.gz file or at least using a file with the first 1k rows to make sure the tests cover some real data variance.

It could be a long-process to find a bug if the dataset is big. For example, you found that `sum(dim1)` suddenly is not correct, with a small dataset you can quickly see the root cause. With more data the fixing process can quickly degrade to shamanism.
I can suggest to have 2 sources for different tests, one set of tests check corner cases and allow to quickly check and fix for regressions.  The other set of sanity tests  can work on a large dataset for additional insurance (we can create them in another PR).

**Comment by begelundmuller** [comment]
> > In runtime/testproject/ad_bids, two rows feels a little fragile for the data. How about either using the real ~5MB adbids.csv.gz file or at least using a file with the first 1k rows to make sure the tests cover some real data variance.
> 
> It could be a long-process to find a bug if the dataset is big. For example, you found that `sum(dim1)` suddenly is not correct, with a small dataset you can quickly see the root cause. With more data the fixing process can quickly degrade to shamanism. I can suggest to have 2 sources for different tests, one set of tests check corner cases and allow to quickly check and fix for regressions. The other set of sanity tests can work on a large dataset for additional insurance (we can create them in another PR).

Okay, but please let's do at least ~25 rows at least so we make sure to cover some variance (e.g. more than one group returned)

File: .gitignore (hunk 1 above)
Comment (egor-ryashin) [reply .gitignore]: fixed

File: runtime/server/metrics_views_test.go (hunk 3 above)
Comment (egor-ryashin) [reply runtime/server/metrics_views_test.go]: fixed

**Comment by begelundmuller** [comment]
> I don't understand why 25? It can return 2 groups with 2 rows, if you need aggregation in both groups then 4 rows suffice.

There's also e.g. time range queries, which should narrow down the dataset. If testing on too small a dataset, I'm just worried we're not surfacing edge cases, e.g. what if a column has some sporadic nulls?

Are you worried about the tests being hard to write? We don't have to separately validate the output data of everything, just make sure the tests cover some real data and appear to give correct output, then we can pin the expected values in the tests so we can at least detect regressions.

**Comment by egor-ryashin** [comment]
Probably you have some test in mind that can spot issues with a large dataset while miss it with the a small dataset, could you provide an example that way I can follow up the approach?

**Comment by egor-ryashin** [comment]
I've added 5mb csv tests.

File: runtime/server/metrics_views.go [hunk 4]
```diff
@@ -346,18 +346,18 @@ func buildFilterClauseForCondition(cond *runtimev1.MetricsViewFilter_Cond, exclu
 	var operatorPrefix string
 	var conditionJoiner string
 	if exclude {
-		operatorPrefix = "NOT"
-		conditionJoiner = "AND"
+		operatorPrefix = " NOT "
```
Comment (AdityaHegde) [inline runtime/server/metrics_views.go:349]: How different is this function vs the one in timeseries_api.go? We should try and maintain a single filter to where clause converter function.

Comment (egor-ryashin) [reply runtime/server/metrics_views.go:349]: There're not different functionally, it's only a matter of time I start refactoring that and merge implementation where possible.

**Review by begelundmuller — APPROVED** [review]
lgtm

**Comment by begelundmuller** [comment]
> Probably you have some test in mind that can spot issues with a large dataset while miss it with the a small dataset, could you provide an example? that way I could follow up the approach

Real datasets will generally surface conditions not considered in tiny datasets, such as groups with no values, pagination when many groups are returned, groups with many null values, wider variety of values (such as negative floats) being handled, join errors leading to huge numbers of rows, etc.

But I think you're right there's an argument to be made for using small targeted datasets for unit testing, and then only using larger datasets in less granular e2e testing. So using small datasets here sounds good to me, we can do large dataset testing later somewhere else.
